### PR TITLE
[MU4] PopupView component

### PR DESCRIPTION
**Pull request by vpereverzev**
This PR introduces a new component PopupView and it's "musescored" version called "StyledPopupView".

Currently, we're not able to use QML Popup inside dockwidgets, which already blocks several of our new features. 
So the main difference between QtQuick Popup and PopupView is that it could be used within the DockWidgets.

Basically, it uses the common interface of QtQuick.Controls with two main properties "backgroundItem" and "contentItem".

Issues that are not covered by this PR:
- we'd need to replace usages of the current StyledPopup with the new component
- Auto-close policy doesn't support "Esc" key-press yet - https://doc.qt.io/qt-5/qml-enumeration.html. The support of this case will be available later.

Peek 2021-02-08 23-48

### CODE REVIEW

**Comment by vpereverzev** [comment]
code style update is coming soon

File: src/appshell/qml/DevTools/Interactive/InteractiveTests.qml [hunk 1]
```diff
@@ -119,10 +119,24 @@ Rectangle {
...
+
         FlatButton {
             width: 200
-            text: "Widget dialog async"
-            onClicked: testModel.openWidgetDialogAsync()
+            text: "Widget dialogg async"
+            onClicked: popup.show()//testModel.openWidgetDialogAsync()
+
+            Component.onCompleted: {
+                var i = 0
```
Comment (RomanPudashkin) [inline src/appshell/qml/DevTools/Interactive/InteractiveTests.qml:138]: Please, remove this code

File: src/appshell/qml/DevTools/Interactive/InteractiveTests.qml [hunk 2]
```diff
@@ -119,10 +119,24 @@ Rectangle {
...
+                height: 300
+            }
+        }
+
         FlatButton {
             width: 200
-            text: "Widget dialog async"
-            onClicked: testModel.openWidgetDialogAsync()
+            text: "Widget dialogg async"
+            onClicked: popup.show()//testModel.openWidgetDialogAsync()
```
Comment (RomanPudashkin) [inline src/appshell/qml/DevTools/Interactive/InteractiveTests.qml:135]: Please, remove this comment

File: src/framework/uicomponents/view/popupview.cpp [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+qreal PopupView::padding() const
+{
+    return m_padding;
+}
+
+void PopupView::setBackgroundItem(QQuickItem* backgroundItem)
+{
+    if (m_backgroundItem == backgroundItem)
```
Comment (RomanPudashkin) [inline src/framework/uicomponents/view/popupview.cpp:85]: Please, add {} here

File: src/framework/uicomponents/view/popupview.cpp [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+        return;
+
+    m_backgroundItem = backgroundItem;
+    emit backgroundItemChanged(m_backgroundItem);
+}
+
+void PopupView::setContentItem(QQuickItem* contentItem)
+{
+    if (m_contentItem == contentItem)
```
Comment (RomanPudashkin) [inline src/framework/uicomponents/view/popupview.cpp:94]: Please, add {} here

File: src/framework/uicomponents/view/popupview.cpp [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    if (QEvent::MouseButtonPress == event->type()) {
+        mousePressEvent(static_cast<QMouseEvent*>(event));
+    } else if (QEvent::MouseButtonRelease == event->type()) {
+        mouseReleaseEvent(static_cast<QMouseEvent*>(event));
+    }
+
+    return QObject::eventFilter(watched, event);
+}
+
+void PopupView::mousePressEvent(QMouseEvent* event)
```
Comment (RomanPudashkin) [inline src/framework/uicomponents/view/popupview.cpp:184]: The implementation of this function duplicates mouseReleaseEvent

File: src/framework/uicomponents/view/popupview.h [hunk 6]
```diff
@@ -0,0 +1,106 @@
+#ifndef POPUPVIEW_H
```
Comment (RomanPudashkin) [inline src/framework/uicomponents/view/popupview.h:1]: Please, add a license to h and cpp and replace POPUPVIEW_H with MU_UICOMPONENTS_POPUPVIEW_H

File: src/framework/uicomponents/view/popupview.h [hunk 7]
```diff
@@ -0,0 +1,106 @@
+#ifndef POPUPVIEW_H
+#define POPUPVIEW_H
+
+#include <QQuickItem>
+#include <QQuickView>
+
+class PopupView : public QQuickItem
```
Comment (RomanPudashkin) [inline src/framework/uicomponents/view/popupview.h:7]: Please, add namespace:

namespace mu::uicomponents {
class PopupView
...
}

File: src/appshell/qml/DevTools/Interactive/InteractiveTests.qml (hunk 2 above)
Comment (Eism) [reply src/appshell/qml/DevTools/Interactive/InteractiveTests.qml:135]: no, needed to revert changes in this file

File: src/framework/uicomponents/qml/MuseScore/UiComponents/StyledPopupView.qml [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    property alias arrowHeight: arrow.height
+
+    property color borderColor: ui.theme.strokeColor
+    property color fillColor: ui.theme.backgroundPrimaryColor
+    readonly property int borderWidth: 1
+
+    closePolicy: PopupView.CloseOnPressOutsideParent
+
+    onAboutToShow: {
+        globalPos = mapToGlobal(x, y)
```
Comment (Eism) [inline src/framework/uicomponents/qml/MuseScore/UiComponents/StyledPopupView.qml:21]: can you move it to c++ class of popupView?

**Review by igorkorsukov — APPROVED** [review]

File: src/framework/uicomponents/view/popupview.cpp (hunk 5 above)
Comment (vpereverzev) [reply src/framework/uicomponents/view/popupview.cpp:184]: not really, it handles 2 different modes of auto-close policy

File: src/appshell/qml/DevTools/Interactive/InteractiveTests.qml (hunk 2 above)
Comment (vpereverzev) [reply src/appshell/qml/DevTools/Interactive/InteractiveTests.qml:135]: yeah, forgotten test code, thx

File: src/framework/uicomponents/qml/MuseScore/UiComponents/StyledPopupView.qml (hunk 8 above)
Comment (vpereverzev) [reply src/framework/uicomponents/qml/MuseScore/UiComponents/StyledPopupView.qml:43]: added in-line comment with an explanation

**Review by RomanPudashkin — APPROVED** [review]

**Pull request merged**